### PR TITLE
Upstream merge 2023-11-01 (part 2)

### PR DESCRIPTION
**Pull request by justsmth**
## Description of changes:

Merging from Upstream from https://github.com/google/boringssl/commit/a905bbb52a7bac5099f2cbee008c6f3eae96218c (Jun 22).

## Call-outs:

See internal document as well as "AWS-LC" notes inserted in some of the commit messages for additions/deviations from the upstream commit.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Comment by justsmth** [comment]
I also built/tested for PPC:
```
❯ ppc-qemu.sh crypto/crypto_test --gtest_filter="AEADTest.*"  
Note: Google Test filter = AEADTest.*
[==========] Running 9 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 9 tests from AEADTest
[ RUN      ] AEADTest.AESCCMLargeAD
[       OK ] AEADTest.AESCCMLargeAD (714 ms)
[ RUN      ] AEADTest.WycheproofAESGCMSIV
[       OK ] AEADTest.WycheproofAESGCMSIV (641 ms)
[ RUN      ] AEADTest.WycheproofAESGCM
[       OK ] AEADTest.WycheproofAESGCM (587 ms)
[ RUN      ] AEADTest.WycheproofChaCha20Poly1305
[       OK ] AEADTest.WycheproofChaCha20Poly1305 (289 ms)
[ RUN      ] AEADTest.WycheproofXChaCha20Poly1305
[       OK ] AEADTest.WycheproofXChaCha20Poly1305 (310 ms)
[ RUN      ] AEADTest.FreeNull
[       OK ] AEADTest.FreeNull (1 ms)
[ RUN      ] AEADTest.AEADAES256GCMDetIVGen
[       OK ] AEADTest.AEADAES256GCMDetIVGen (0 ms)
[ RUN      ] AEADTest.TestGCMSIV128Change16Alignment
...
[       OK ] AEADTest.TestGCMSIV256Change16Alignment (5 ms)
[----------] 9 tests from AEADTest (2555 ms total)

[----------] Global test environment tear-down
[==========] 9 tests from 1 test suite ran. (2562 ms total)
[  PASSED  ] 9 tests.
```

File: include/openssl/target.h [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+#define OPENSSL_PNACL
+#elif defined(__wasm__)
+#define OPENSSL_32_BIT
+#elif defined(__asmjs__)
+#define OPENSSL_32_BIT
+#elif defined(__myriad2__)
+#define OPENSSL_32_BIT
+#else
+// Note BoringSSL only supports standard 32-bit and 64-bit two's-complement,
+// little-endian architectures. Functions will not produce the correct answer
```
[the remark this answers is not in the document]
Comment (nebeid) [reply include/openssl/target.h:73]: Maybe we can merge this cherry-picked commit and change the comment in another commit.

File: include/openssl/asm_base.h [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+#define OPENSSL_ASM_INCOMPATIBLE
+#endif
+#endif
+
+#if defined(OPENSSL_ASM_INCOMPATIBLE)
+#undef OPENSSL_ASM_INCOMPATIBLE
+#if !defined(OPENSSL_NO_ASM)
+#define OPENSSL_NO_ASM
+#endif
+#endif  // OPENSSL_ASM_INCOMPATIBLE
```
Comment (nebeid) [inline include/openssl/asm_base.h:41]: - This section is at the end of target.h, why is it repeated here?
- If HWASAN is a new define, should we put its motivation in the commit message under an "AWS-LC:" header?

File: include/openssl/target.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (justsmth) [reply include/openssl/target.h]: I updated this comment in the latest commit.

File: include/openssl/asm_base.h (hunk 2 above)
Comment (justsmth) [reply include/openssl/asm_base.h]: * Removed section from here.
* Updated [commit comment](https://github.com/aws/aws-lc/pull/1278/commits/63f51bcd77f7c19cf5ca588a942d54680bb3c1da) to mention adding `OPENSSL_HWASAN`.

**Review by skmcgrail — APPROVED** [review]

**Review by nebeid — APPROVED** [review]

**Pull request merged**